File: Binary_Tree/107.Binary_Tree_Level_Order_Traversal_II/Python/107.Binary_Tree_Level_Order_Traversal_II.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def levelOrderBottom(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: List[List[int]]
        """
        #============================================#
        # Breath-first search based traversal method #
        #============================================#

        ############
        #Initialize
        ##### Record queue #####
        record_queue = []

        ##### Result array #####
        res_arry = []


        ##########################################
        #Breath-first search based loop traversal
        if (not root): #Issue/Boundary-case handling
            return res_arry

        record_queue.append(root) #Update/Record

        while (record_queue):
            len_queue = len(record_queue) #Length of queue array

            record_arry = [] #Record array

            for queue_idx in range(len_queue):
                record_root = record_queue.pop(0) #Record root

                record_arry.append(record_root.val) #Keep updating/recording

                ##### Check if the root left-half node existed or not #####
                if (record_root.left):
                    record_queue.append(record_root.left) #Keep updating/recording
                else:
                    pass

                ##### Check if the root right-half node existed or not #####
                if (record_root.right):
                    record_queue.append(record_root.right) #Keep updating/recording
                else:
                    pass

            res_arry.append(record_arry) #Keep updating/recording

        return res_arry[:: (-1)]

File: Binary_Tree/107.Binary_Tree_Level_Order_Traversal_II/C++/107.Binary_Tree_Level_Order_Traversal_II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode * root) 
    {
        //=========================================//
        // Input type:                             //
        //   - TreeNode * (root tree-node)         //
        // Return type:                            //
        //   - vector<vector<int>> (result vector) //
        //=========================================//

        /*Initialize*/
        ///// Record queue /////
        queue<TreeNode * > record_queue;

        ///// Result vector /////
        vector<vector<int>> res_vect;


        /*Breath-first search based loop traversal*/
        if (root == nullptr) { return res_vect; } //Issue, Boundary-case handling

        record_queue.emplace(root); //Update, Record

        while (! record_queue.empty()) //Cycled-times
        {
            vector<int> record_vect; //Record vector

            for (int queue_idx = record_queue.size(); (queue_idx > 0); (queue_idx--)) //Queue-size
            {
                TreeNode * record_root = record_queue.front(); record_queue.pop(); //Keep updating, popped

                record_vect.emplace_back((record_root->val)); //Keep updating, recording

                ///// Check if the root left-half node existed or not /////
                if ((record_root->left) != nullptr) { record_queue.emplace((record_root->left)); } //Keep updating, recording
                else { ; }

                ///// Check if the root right-half node existed or not /////
                if ((record_root->right) != nullptr) { record_queue.emplace((record_root->right)); } //Keep updating, recording
                else { ; }

            } //Queue-size

            res_vect.push_back(record_vect); //Keep updating, recording

        } //Cycled-times

        reverse(res_vect.begin(), res_vect.end()); //Reversed operations

        return res_vect;
    }
};
